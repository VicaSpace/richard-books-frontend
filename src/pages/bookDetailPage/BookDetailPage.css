.book-detail-container {
  padding: 2.75rem 1.5rem;
}

.book-detail-op-bar {
  display: flex;
  justify-content: space-between;
}

.book-detail-back {
  size: 5rem;
}

.book-detail-bookmark-yellow {
  color: #FBE207;
}

.book-detail-cover {
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 1rem 0;
}

.book-detail-info {
  font-family: 'Open Sans', sans-serif;
}

.book-detail-info p {
  margin: 0;
}

.book-detail-name {
  font-weight: bolder;
  font-size: 1.5rem;
  text-align: center;
  padding: 0.5rem;
}

.book-detail-author {
  text-align: center;
  color: #9797AA;
  padding-bottom: 0.5rem;
}

.book-detail-rating {
  padding-bottom: 0.5rem;
  display: flex;
  justify-content: center;
  align-items: center;
}

.book-detail-star {
  color: #FFD335;
  margin-right: 0.2rem;
}

.book-detail-emp {
  margin-left: 0.5rem;
  margin-right: 0.2rem;
}

.book-detail-norm {
  color: #9797AA;
  font-size: 0.9rem;
}

.book-detail-description {
  font-size: 0.8rem;
}

.book-detail-button button {
  width: 100%;
  margin: 1rem 0;
  padding: 0.8rem 0;
  color: white;
  background: black;
  font-size: 1rem;
  border-radius: 1rem;
  border: 0;
}
