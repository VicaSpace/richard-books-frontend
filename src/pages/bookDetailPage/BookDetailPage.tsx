import React, { useEffect, useState } from 'react'
import { RiBookmarkFill } from 'react-icons/ri'
import { useNavigate, useParams } from 'react-router'
import { Book } from '../../App'
import { IoIosArrowBack } from 'react-icons/io'
import './BookDetailPage.css'
import { AiFillStar } from 'react-icons/ai'

const BookDetailPage: React.FC<{ books: Book[] }> = ({ books }) => {
  const { id } = useParams<string>()
  const [book, setBook] = useState<Book | undefined>()
  const [isBookmarked, setBookmark] = useState<boolean>(false)
  const navigate = useNavigate()

  useEffect(() => {
    if (typeof id === 'undefined') {
      return
    }
    const selectedBook = books.filter((e) => e.id === parseInt(id, 10))[0]
    setBook(selectedBook)
  }, [books])

  useEffect(() => {
    if (typeof book === 'undefined') {
      return
    }
    const strBookmarkIds = localStorage.getItem('bookmark') ?? '[]'
    const bookmarkIds: number[] = JSON.parse(strBookmarkIds)
    if (bookmarkIds.includes(book.id)) {
      setBookmark(true)
    }
  }, [book])

  const onBookmarkIconClickHandler = (): void => {
    if (typeof book === 'undefined') {
      return
    }
    setBookmark(!isBookmarked)
    const strBookmarkIds = localStorage.getItem('bookmark') ?? '[]'
    const bookmarkIds: number[] = JSON.parse(strBookmarkIds)
    if (bookmarkIds.includes(book.id)) {
      const newBookmarkIds = bookmarkIds.filter((e) => e !== book.id)
      localStorage.setItem('bookmark', JSON.stringify(newBookmarkIds))
    } else {
      bookmarkIds.push(book.id)
      localStorage.setItem('bookmark', JSON.stringify(bookmarkIds))
    }
  }

  return (
    <>
      {typeof book === 'undefined'
        ? (
        <>Error</>
          )
        : (
        <div className="book-detail-container">
          <div className="book-detail-op-bar">
            <IoIosArrowBack
              className="book-detail-back"
              onClick={() => navigate(-1)}
              size="1.5rem"
            />
            <div className="book-detail-bookmark-container">
              {isBookmarked
                ? (
                <RiBookmarkFill
                  className="book-detail-bookmark-yellow"
                  onClick={onBookmarkIconClickHandler}
                  size="1.5rem"
                />
                  )
                : (
                <RiBookmarkFill
                  className="book-detail-bookmark"
                  onClick={onBookmarkIconClickHandler}
                  size="1.5rem"
                />
                  )}
            </div>
          </div>
          <div className="book-detail-cover">
            <img src={book.img} alt={book.name} />
          </div>
          <div className="book-detail-info">
            <p className="book-detail-name">{book.name}</p>
            <p className="book-detail-author">{book.Author.name}</p>
            <div className="book-detail-rating">
              <AiFillStar className='book-detail-star' />
              <AiFillStar className='book-detail-star' />
              <AiFillStar className='book-detail-star' />
              <AiFillStar className='book-detail-star' />
              <AiFillStar className='book-detail-star' />
              <span className='book-detail-emp'>5.0</span>
              <span className='book-detail-norm'>/ 5.0</span>
            </div>
            <p className="book-detail-description">{book.description}</p>
          </div>
          <div className="book-detail-button">
            <button>Buy Now for ${book.price}</button>
          </div>
        </div>
          )}
    </>
  )
}

export default BookDetailPage
