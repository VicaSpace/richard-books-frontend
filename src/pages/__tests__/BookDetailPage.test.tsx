import React from 'react'
import { render } from '@testing-library/react'
import BookDetailPage from '../bookDetailPage/BookDetailPage'
import { MemoryRouter, Route, Routes } from 'react-router-dom'

const books = [
  {
    id: 1,
    name: 'Book 1',
    img: 'img src',
    genre: 'Children',
    description: 'Description',
    price: 5,
    Author: {
      id: 1,
      name: 'John',
      img: "John's image"
    }
  },
  {
    id: 2,
    name: 'Book 2',
    img: 'img src',
    genre: 'Children',
    description: 'Description',
    price: 5,
    Author: {
      id: 1,
      name: 'John',
      img: "John's image"
    }
  }
]

describe('test book detail page', () => {
  it('should render page correctly', () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/books/1']}>
        <Routes>
          <Route path='/books/:id' element={<BookDetailPage books={books} />} />
        </Routes>
      </MemoryRouter>
    )
    expect(container).toMatchSnapshot()
  })
})
