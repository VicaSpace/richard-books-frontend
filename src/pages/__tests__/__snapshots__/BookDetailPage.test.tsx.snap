// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test book detail page should render page correctly 1`] = `
<div>
  <div
    class="book-detail-container"
  >
    <div
      class="book-detail-op-bar"
    >
      <svg
        class="book-detail-back"
        fill="currentColor"
        height="1.5rem"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1.5rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M217.9 256L345 129c9.4-9.4 9.4-24.6 0-33.9-9.4-9.4-24.6-9.3-34 0L167 239c-9.1 9.1-9.3 23.7-.7 33.1L310.9 417c4.7 4.7 10.9 7 17 7s12.3-2.3 17-7c9.4-9.4 9.4-24.6 0-33.9L217.9 256z"
        />
      </svg>
      <div
        class="book-detail-bookmark-container"
      >
        <svg
          class="book-detail-bookmark"
          fill="currentColor"
          height="1.5rem"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M5 2h14a1 1 0 0 1 1 1v19.143a.5.5 0 0 1-.766.424L12 18.03l-7.234 4.536A.5.5 0 0 1 4 22.143V3a1 1 0 0 1 1-1z"
            />
          </g>
        </svg>
      </div>
    </div>
    <div
      class="book-detail-cover"
    >
      <img
        alt="Book 1"
        src="img src"
      />
    </div>
    <div
      class="book-detail-info"
    >
      <p
        class="book-detail-name"
      >
        Book 1
      </p>
      <p
        class="book-detail-author"
      >
        John
      </p>
      <div
        class="book-detail-rating"
      >
        <svg
          class="book-detail-star"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
          />
        </svg>
        <svg
          class="book-detail-star"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
          />
        </svg>
        <svg
          class="book-detail-star"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
          />
        </svg>
        <svg
          class="book-detail-star"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
          />
        </svg>
        <svg
          class="book-detail-star"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
          />
        </svg>
        <span
          class="book-detail-emp"
        >
          5.0
        </span>
        <span
          class="book-detail-norm"
        >
          / 5.0
        </span>
      </div>
      <p
        class="book-detail-description"
      >
        Description
      </p>
    </div>
    <div
      class="book-detail-button"
    >
      <button>
        Buy Now for $
        5
      </button>
    </div>
  </div>
</div>
`;
